wdl-lint/src/rules/runtime_section_keys.rs

```diff
@@ -0,0 +1,502 @@
+    ///
+    /// This is used to determine the missing keys at the end of the task
+    /// callback.
+    encountered_keys: Vec<String>,
+    /// Any unknown keys encountered.
+    non_reserved_keys: HashMap<String, Span>,
+}
+
+impl Rule for RuntimeSectionKeysRule {
+    fn id(&self) -> &'static str {
+        ID
+    }
+
+    fn description(&self) -> &'static str {
+        "Ensures that `runtime` sections have the appropriate keys."
+    }
+
+    fn explanation(&self) -> &'static str {
+        "The behavior of this rule is different depending on the WDL version:
+        
+        For WDL v1.0 documents, the `docker` and `memory` keys are recommended, but the inclusion \
+         of any number of other keys is permitted.
+        
+        For WDL v1.1 documents, 
+        
+        - A list of mandatory, reserved keywords will be recommended for inclusion if they are not \
+         present. Here, 'mandatory' refers to the requirement that all execution engines support \
+         this key—not that the key must be present in the `runtime` section.
+        - Optional, reserved \"hint\" keys are also permitted but not flagged when they are \
+         missing (as their support in execution engines is not guaranteed).
+        - The WDL v1.1 specification deprecates the inclusion of non-reserved keys in a  `runtime` \
+         section. As such, any non-reserved keys will be flagged for removal.
+         
+         For WDL v1.2 documents and later, this rule does not evaluate because `runtime` sections \
+         were deprecated in this version."
+    }
+
+    fn tags(&self) -> crate::TagSet {
+        TagSet::new(&[Tag::Completeness, Tag::Deprecated])
+    }
+}
+
+/// A utility method to parse the recommended keys from a static set of runtime
+/// keys from either WDL v1.0 or WDL v1.1.
+fn recommended_keys<'a, 'k>(
+    keys: &'a HashMap<&'k str, KeyKind>,
+) -> impl Iterator<Item = (&'k str, &'a KeyKind)> {
+    keys.iter()
+        .filter(|(_, kind)| kind.is_recommended())
+        .map(|(key, kind)| (*key, kind))
+}
+
+impl Visitor for RuntimeSectionKeysRule {
+    type State = Diagnostics;
+
+    fn document(
+        &mut self,
+        _: &mut Self::State,
+        reason: wdl_ast::VisitReason,
+        _: &wdl_ast::Document,
+        version: Option<SupportedVersion>,
+    ) {
+        if reason == VisitReason::Exit {
+            return;
+        }
+
+        // Reset the visitor upon document entry.
+        *self = Default::default();
+
+        // NOTE: this rule is dependent on the version of the WDL document.
+        self.version = Some(version);
+    }
+
+    fn task_definition(
+        &mut self,
+        state: &mut Self::State,
+        reason: VisitReason,
+        _: &TaskDefinition,
+    ) {
+        match reason {
+            VisitReason::Enter => {
+                self.current_runtime_span = None;
+                self.encountered_keys = Default::default();
```

Comment:
The fix for the issue you were seeing:

```suggestion
                self.encountered_keys.clear();
                self.non_reserved_keys.clear();
```

Reply:
There is a distinction between assigning `Default::default()` and calling `clear()`: the latter will preserve the capacity. 

While it's a minor thing, I think reducing allocations when it's trivial to do so would be ideal. That said, I don't think it's worth holding up the PR unless you'd like to fix.